### PR TITLE
[FIRRTL] Implement SFC's RemoveCHIRRTL pass

### PR DESCRIPTION
**Pull request by youngar**
This adds an implementation of the `RemoveCHIRRTL` pass called
`InferMemories`.  This pass takes the CHIRRTL memory operations, `seqmem`
and `combmem`, and transforms them into standard FIRRTL `mem`
operations.

### CODE REVIEW

**Review by lattner — COMMENTED** [review]
This patch is looking really really great, I looked at all of it except the new pass.

Could you break this into three different patches:

1) introduce the changes to the IR, e.g. introducing MemoryType and renaming the ops.
2) Change the parser behavior.
3) Introduce the new pass

That will help with review, thank you!

File: include/circt/Dialect/FIRRTL/Passes.td [hunk 1]
```diff
@@ -100,6 +100,40 @@ def ExpandWhens : Pass<"firrtl-expand-whens", "firrtl::FModuleOp"> {
   let constructor = "circt::firrtl::createExpandWhensPass()";
 }
 
+def InferMemories : Pass<"firrtl-infer-memories", "firrtl::FModuleOp"> {
+  let summary = "Infer the memory ports of SeqMem and CombMem";
+  let description = [{
+      This pass finds the CHIRRTL behavioural memories and their ports, and
```
[the remark this answers is not in the document]
Comment (seldridge) [reply include/circt/Dialect/FIRRTL/Passes.td:106]: 🇨🇦

File: include/circt/Dialect/FIRRTL/Passes.td [hunk 2]
```diff
@@ -100,6 +100,39 @@ def ExpandWhens : Pass<"firrtl-expand-whens", "firrtl::FModuleOp"> {
   let constructor = "circt::firrtl::createExpandWhensPass()";
 }
 
+def InferMemories : Pass<"firrtl-infer-memories", "firrtl::FModuleOp"> {
```
Comment (youngar) [inline include/circt/Dialect/FIRRTL/Passes.td:103]: Any suggestion for a better name?  In our previous notes we were calling it "LowerBehavioralMemories" which I like better, although we have been calling them CHIRRTL memories more often than behavioral memories.

Comment (youngar) [reply include/circt/Dialect/FIRRTL/Passes.td:103]: I might end up folding SFC's `InferReadWrite` pass functionality into this, so the name is probably ok.  `InferReadWrite` combines separate `Read` and `Write` ports into a single `ReadWrite` port)

File: lib/Dialect/FIRRTL/Transforms/InferMemories.cpp [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+  void visitExpr(SubaccessOp op);
+  void visitExpr(SubfieldOp op);
+  void visitExpr(SubindexOp op);
+  void visitStmt(ConnectOp op);
+  void visitStmt(PartialConnectOp op);
+  void visitUnhandledOp(Operation *op);
+
+  /// Get a the constant 0.  This constant is inserted at the beginning of the
+  /// module.
+  Value getConstZero() {
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/InferMemories.cpp:50]: just getConst(val)?

File: lib/Dialect/FIRRTL/Transforms/InferMemories.cpp [hunk 4]
```diff
@@ -0,0 +1,731 @@
...
+  // operation we store how it is used into a global hashtable for later use.
+  // This records how both the MemoryPort and Subfield operations are used.
+  struct StackElement {
+    Value value;
+    Value::use_iterator iterator;
+    MemDirAttr mode;
+  };
+
+  SmallVector<StackElement> stack;
+  stack.push_back(
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/InferMemories.cpp:239]: emplace_back?

**Review by darthscsi — APPROVED** [review]
Looks good.  I'm looking forwards to trying on Monday.

File: include/circt/Dialect/FIRRTL/Passes.td (hunk 2 above)
Comment (lattner) [reply include/circt/Dialect/FIRRTL/Passes.td:103]: How about `lower-chirrtl` ?

File: lib/Dialect/FIRRTL/Transforms/InferMemories.cpp [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+                                    Value value) {
+  forEachLeaf(builder, value, [&](Value leaf) {
+    builder.create<ConnectOp>(leaf, getInvalid(leaf.getType()));
+  });
+}
+
+/// Combines two MemDirAttrs into a new one which maintains the union of the
+/// Read and Write options.
+static MemDirAttr mergeDirection(MemDirAttr a, MemDirAttr b) {
+  if (a == MemDirAttr::ReadWrite || b == MemDirAttr::ReadWrite)
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/InferMemories.cpp:180]: Would it makes sense to define `MemDirAttr` as an enum of "1, 2, 3" so we can `or` these together?

I think this helper can get hoisted up to a header or something if it is small

File: tools/firtool/firtool.cpp [hunk 6]
```diff
@@ -230,13 +235,20 @@ processBuffer(MLIRContext &context, TimingScope &ts, llvm::SourceMgr &sourceMgr,
...
+        firrtl::createInferMemoriesPass());
+
   // Width inference creates canonicalization opportunities.
   if (inferWidths)
     pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInferWidthsPass());
 
   if (inferResets)
     pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInferResetsPass());
 
+  if (blackBoxMemory)
```
Comment (lattner) [inline tools/firtool/firtool.cpp:249]: Why move this?  Can you split any unrelated changes to a separate patch?

**Review by lattner — APPROVED** [review]
Looking nice to me.

File: tools/firtool/firtool.cpp (hunk 6 above)
Comment (youngar) [reply tools/firtool/firtool.cpp]: I added a memory integration test that is making sure that this stuff was all working together, and I found that BlackBoxMemory isn't working anymore, and needs to run before LowerTypes.  I can move this + the integration test to a new PR.

File: lib/Dialect/FIRRTL/Transforms/InferMemories.cpp (hunk 5 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/InferMemories.cpp]: Yeah, this makes sense. They were already defined as "1, 2, 3", so I just added `operation|` and `operator|=` in `FIRRTLOps.h` where we tend to put a lot of this supporting stuff.

File: lib/Dialect/FIRRTL/Transforms/InferMemories.cpp [hunk 7]
```diff
@@ -0,0 +1,731 @@
...
+              [&](Value leaf) { builder.create<ConnectOp>(leaf, value); });
+}
+
+/// Connect each leaf of an aggregate type to invalid.  This does not support
+/// aggregates with flip types.
+void InferMemoriesPass::emitInvalid(ImplicitLocOpBuilder &builder,
+                                    Value value) {
+  forEachLeaf(builder, value, [&](Value leaf) {
+    builder.create<ConnectOp>(leaf, getInvalid(leaf.getType()));
+  });
```
[the remark this answers is not in the document]
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/InferMemories.cpp]: Yeah that is a lot simpler and I was able to remove some of the helper functions.

File: lib/Dialect/FIRRTL/Transforms/InferMemories.cpp [hunk 8]
```diff
@@ -0,0 +1,731 @@
...
+  MemDirAttr inferMemoryPortKind(MemoryPortOp memPort);
+
+  void replaceMem(Operation *op, StringRef name, bool isSequential, RUWAttr ruw,
+                  ArrayAttr annotations);
+
+  template <typename OpType, typename... T>
+  void cloneSubindexOpForMemory(OpType op, Value input, T... operands);
+
+  void emitPartialConnectMask(
+      ImplicitLocOpBuilder builder, Type destType, Type srcType,
```
[the remark this answers is not in the document]
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/InferMemories.cpp]: Fixed, thanks!

File: lib/Dialect/FIRRTL/Transforms/InferMemories.cpp [hunk 9]
```diff
@@ -0,0 +1,731 @@
...
+  void visitStmt(PartialConnectOp op);
+  void visitUnhandledOp(Operation *op);
+
+  /// Get a the constant 0.  This constant is inserted at the beginning of the
+  /// module.
+  Value getConstZero() {
+    if (!constZero) {
+      auto builder = OpBuilder::atBlockBegin(getOperation().getBodyBlock());
+      auto u1Type = UIntType::get(builder.getContext(), /*width*/ 1);
+      constZero = builder.create<ConstantOp>(builder.getUnknownLoc(), u1Type,
```
[the remark this answers is not in the document]
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/InferMemories.cpp]: Fixed!

File: lib/Dialect/FIRRTL/Transforms/InferMemories.cpp (hunk 4 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/InferMemories.cpp]: Fixed, thanks!

File: tools/firtool/firtool.cpp (hunk 6 above)
Comment (lattner) [reply tools/firtool/firtool.cpp]: cool thanks
